stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -616,166 +335,73 @@ private void releaseChildNodeRemovedEvent(ChildNodeRemovedEvent childNodeRemoved
+
+      ElementInfo oldElementInfo = mDocument.getShadowDOM().getElementInfo(element);
+      removedEvent.parentNodeId = mDocument.getObjectIdMapper().getIdForObject(oldElementInfo.parentElement);
+      removedEvent.nodeId = mDocument.getObjectIdMapper().getIdForObject(element);
       mPeerManager.sendNotificationToPeers("DOM.childNodeRemoved", removedEvent);
+
       releaseChildNodeRemovedEvent(removedEvent);
     }
-  }
 
-  private final class DOMObjectIdMapper extends ObjectIdMapper {
-    @Override
-    protected void onMapped(Object object, int id) {
-      mDOMProvider.verifyThreadAccess();
+    public void onChildNodeInserted(Object element, int parentNodeId, int previousNodeId, DOMView view, Accumulator<Object> insertedElements) {
+      ChildNodeInsertedEvent insertedEvent = acquireChildNodeInsertedEvent();
 
-      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
-      descriptor.hook(object);
-    }
+      insertedEvent.parentNodeId = parentNodeId;
+      insertedEvent.previousNodeId = previousNodeId;
+      insertedEvent.node = createNodeForElement(view, element);
 
-    @Override
-    protected void onUnmapped(Object object, int id) {
-      mDOMProvider.verifyThreadAccess();
+      insertedElements.store(element);
+      mPeerManager.sendNotificationToPeers("DOM.childNodeInserted", insertedEvent);
 
-      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
-      descriptor.unhook(object);
+      releaseChildNodeInsertedEvent(insertedEvent);
     }
   }
 
-  private final class ProviderListener implements DOMProvider.Listener {
-    @Override
-    public void onPossiblyChanged() {
-      updateTree();
-    }
-
+  private final class PeerManagerListener extends PeersRegisteredListener {
     @Override
-    public void onAttributeModified(Object element, String name, String value) {
-      mDOMProvider.verifyThreadAccess();
+    protected synchronized void onFirstPeerRegistered() {
+      // We don't call mDOMProvider.setListener() until the first DOM.getDocument request.
 
-      AttributeModifiedEvent message = new AttributeModifiedEvent();
-      message.nodeId = mObjectIdMapper.getIdForObject(element);
-      message.name = name;
-      message.value = value;
-      mPeerManager.sendNotificationToPeers("DOM.attributeModified", message);
+      mDocument.createDomProviderIfNoneExists();
     }
 
     @Override
-    public void onAttributeRemoved(Object element, String name) {
-      mDOMProvider.verifyThreadAccess();
-
-      AttributeRemovedEvent message = new AttributeRemovedEvent();
-      message.nodeId = mObjectIdMapper.getIdForObject(element);
-      message.name = name;
-      mPeerManager.sendNotificationToPeers("DOM.attributeRemoved", message);
-    }
-
-    @Override
-    public void onInspectRequested(Object element) {
-      mDOMProvider.verifyThreadAccess();
+    protected synchronized void onLastPeerUnregistered() {
+      mDocument.cleanUp();
```

Comment:
Same as above.  DOM is doing the state management for Document, even though CSS also uses it.